### PR TITLE
Update the networking dsl words to support IPv6

### PR DESCRIPTION
**Pull request by TrentHouliston**
This pulls the IPv6 changes out of #70 so they can be reviewed separately of the updated tests.
The new tests in #70 verify this in much more detail.
The existing tests were modified enough that they work for their old functionality

### CODE REVIEW

File: src/util/network/if_number_from_address.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#include "get_interfaces.hpp"
+
+namespace NUClear {
+namespace util {
+    namespace network {
+
+        unsigned int if_number_from_address(const sockaddr_in6& ipv6) {
+
+            // If the socket is referring to the any address, return 0 to use the default interface
+            if (std::all_of(ipv6.sin6_addr.s6_addr, ipv6.sin6_addr.s6_addr + 16, [](unsigned char i) {
```
Comment (Bidski) [inline src/util/network/if_number_from_address.cpp:34]: Or can you use `std::begin`/`std::end` here?
```suggestion
            if (std::all_of(ipv6.sin6_addr.s6_addr, ipv6.sin6_addr.s6_addr + sizeof(in6_addr), [](unsigned char i) {
```

File: src/util/network/resolve.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            }
+
+            // Check if we have any addresses to work with
+            if (servinfo_ptr == nullptr) {
+                throw std::runtime_error("Unable to find an address for " + address + ":" + std::to_string(port));
+            }
+
+            std::unique_ptr<addrinfo, void (*)(addrinfo*)> servinfo(servinfo_ptr, ::freeaddrinfo);
+
+            // Empty sock_t struct
```
Comment (Bidski) [inline src/util/network/resolve.cpp:54]: Very descriptive

File: src/util/network/resolve.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+            // The list is actually a linked list of valid addresses
+            // The address we choose is in the following priority, IPv4, IPv6, Other
+            for (addrinfo* p = servinfo.get(); p != nullptr; p = p->ai_next) {
+
+                // If we find an IPv4 address, prefer that
+                if (servinfo->ai_family == AF_INET) {
+
+                    // Clear and set our struct
+                    std::memcpy(&target, servinfo->ai_addr, servinfo->ai_addrlen);
```
Comment (Bidski) [inline src/util/network/resolve.cpp:65]:
```suggestion
                    // Set our target IPv4 address
                    std::memcpy(&target, servinfo->ai_addr, servinfo->ai_addrlen);
```

File: src/util/network/resolve.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+                    // We prefer IPv4 so use it and stop looking
+                    return target;
+                }
+
+                // If we find an IPv6 address, hold the first one in case we don't find an IPv4 address
+                if (servinfo->ai_family == AF_INET6) {
+
+                    // Clear and set our struct
```
Comment (Bidski) [inline src/util/network/resolve.cpp:74]:
```suggestion
                    // // Set our target IPv6 address
```

File: src/util/platform.hpp [hunk 5]
```diff
@@ -78,6 +78,16 @@
     // Whoever thought this was a good idea was a terrible person
     #undef ERROR
 
+    // Make the windows shutdown functions look like the posix ones
+    #define SHUT_RD   SD_RECEIVE
+    #define SHUT_WR   SD_RECEIVE
```
Comment (Bidski) [inline src/util/platform.hpp:83]: ??
```suggestion
    #define SHUT_WR   SD_SEND
```

File: tests/dsl/UDPBroadcast.cpp [hunk 6]
```diff
@@ -66,22 +66,16 @@ class TestReactor : public NUClear::Reactor {
             // Get all the network interfaces
             auto interfaces = NUClear::util::network::get_interfaces();
 
-            std::vector<in_addr_t> addresses;
-
+            std::vector<std::string> addresses;
             for (auto& iface : interfaces) {
-                // We send on broadcast addresses and we don't want loopback or point to point
+                // We send on multicast capable addresses
```
Comment (Bidski) [inline tests/dsl/UDPBroadcast.cpp:71]: Is this `multicast` or `broadcast`?

File: tests/dsl/UDPMulticastKnownPort.cpp [hunk 7]
```diff
@@ -59,31 +59,9 @@ class TestReactor : public NUClear::Reactor {
         // Test with known port
         on<Trigger<Message>>().then([this] {
             // Get all the network interfaces
```
Comment (Bidski) [inline tests/dsl/UDPMulticastKnownPort.cpp:61]:
```suggestion
```

File: tests/dsl/UDPMulticastKnownPort.cpp [hunk 8]
```diff
@@ -59,31 +59,9 @@ class TestReactor : public NUClear::Reactor {
...
-            }
-
-            num_addresses = addresses.size();
 
-            for (auto& ad : addresses) {
-
-                // Send our message to that broadcast address
-                emit<Scope::UDP>(std::make_unique<std::string>(TEST_STRING), MULTICAST_ADDRESS, PORT, ad, in_port_t(0));
-            }
+            // Send our message to that broadcast address
```
Comment (Bidski) [inline tests/dsl/UDPMulticastKnownPort.cpp:63]: What address?

File: tests/dsl/UDPMulticastUnknownPort.cpp [hunk 9]
```diff
@@ -60,35 +60,8 @@ class TestReactor : public NUClear::Reactor {
         // Test with port given to us
         on<Trigger<Message>>().then([this, bound_port] {
             // Get all the network interfaces
```
Comment (Bidski) [inline tests/dsl/UDPMulticastUnknownPort.cpp:62]:
```suggestion
```

File: tests/dsl/UDPMulticastUnknownPort.cpp [hunk 10]
```diff
@@ -60,35 +60,8 @@ class TestReactor : public NUClear::Reactor {
...
-            for (auto& ad : addresses) {
-
-                // Send our message to that broadcast address
-                emit<Scope::UDP>(std::make_unique<std::string>(TEST_STRING),
-                                 MULTICAST_ADDRESS,
-                                 bound_port,
-                                 ad,
-                                 in_port_t(0));
-            }
+            // Send our message to that broadcast address
```
Comment (Bidski) [inline tests/dsl/UDPMulticastUnknownPort.cpp:63]: What address?

File: src/dsl/word/UDP.hpp [hunk 11]
```diff
@@ -105,332 +134,403 @@ namespace dsl {
...
                 }
 
-                // The address we will be binding to
-                sockaddr_in address{};
-                std::memset(&address, 0, sizeof(sockaddr_in));
-                address.sin_family      = AF_INET;
-                address.sin_port        = htons(port);
-                address.sin_addr.s_addr = htonl(INADDR_ANY);
+                int yes = 1;
+                // Include struct in the message "ancillary" control data
```
Comment (Bidski) [inline src/dsl/word/UDP.hpp:186]:
```suggestion
                // Include struct in the message "ancillary" control data
                int yes = 1;
```

**Review by Bidski — CHANGES_REQUESTED** [review]

File: tests/dsl/UDPMulticastUnknownPort.cpp (hunk 10 above)
Comment (TrentHouliston) [reply tests/dsl/UDPMulticastUnknownPort.cpp:63]: Tests are being refactored in #70

File: tests/dsl/UDPMulticastUnknownPort.cpp (hunk 9 above)
Comment (TrentHouliston) [reply tests/dsl/UDPMulticastUnknownPort.cpp:62]: Tests are being refactored in #70

File: tests/dsl/UDPMulticastKnownPort.cpp (hunk 8 above)
Comment (TrentHouliston) [reply tests/dsl/UDPMulticastKnownPort.cpp:63]: Tests are being refactored in #70

File: tests/dsl/UDPMulticastKnownPort.cpp (hunk 7 above)
Comment (TrentHouliston) [reply tests/dsl/UDPMulticastKnownPort.cpp:61]: Tests are being refactored in #70

File: tests/dsl/UDPBroadcast.cpp (hunk 6 above)
Comment (TrentHouliston) [reply tests/dsl/UDPBroadcast.cpp:71]: Tests are being refactored in #70

File: src/dsl/word/emit/UDP.hpp [hunk 12]
```diff
@@ -46,75 +47,102 @@ namespace dsl {
...
              * @param data      the data to emit
-             * @param to_addr   a string or host endian integer specifying the ip to send the packet to
+             * @param to_addr   a string specifying the address to send this packet to
              * @param to_port   the port to send this packet to in host endian
-             * @param from_addr Optional.  A string or host endian integer specifying the local ip to send the packet
-             *                  from.  Defaults to INADDR_ANY.
-             * @param from_port Optional.  The port to send this from to in host endian or 0 to automatically choose a
-             *                  port. Defaults to 0.
+             * @param from_addr Optional. The address to send this from or "" to automatically choose an address.
+             * @param from_port Optional. The port to send this from in host endian or 0 to automatically choose a port.
```
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply src/dsl/word/emit/UDP.hpp:53]: The arguments need defaulting here so you can only provide some of them, e.g. you must provide to addr and port, but you can then only provide a from addr, not a from port. Or provide all 4

File: src/dsl/word/emit/UDP.hpp [hunk 13]
```diff
@@ -46,75 +47,102 @@ namespace dsl {
...
+                    const util::network::sock_t remote = util::network::resolve(to_addr, to_port);
+                    const bool multicast =
+                        remote.sock.sa_family == AF_INET    ? ((remote.ipv4.sin_addr.s_addr >> 28) == 0xE)
+                        : remote.sock.sa_family == AF_INET6 ? ((remote.ipv6.sin6_addr.s6_addr[0] & 0xFF) == 0xFF)
+                                                            : false;
+
+                    // If we are not provided a from address, use any from address
+                    util::network::sock_t local{};
+                    if (from_addr.empty()) {
+                        local = remote;
```
Comment (TrentHouliston) [inline src/dsl/word/emit/UDP.hpp:76]:
```suggestion
                        // By default have the settings of local match remote (except address and port)
                        local = remote;
```

File: src/dsl/word/emit/UDP.hpp [hunk 14]
```diff
@@ -46,75 +47,102 @@ namespace dsl {
...
                             throw std::system_error(network_errno,
                                                     std::system_category(),
-                                                    "We were unable to use the requested interface for multicast");
+                                                    "We were unable to bind the UDP socket to the port");
                         }
                     }
 
-                    // This isn't the greatest code, but lets assume our users don't send broadcasts they don't mean
-                    // to...
+                    // Assume that if the user is sending a broadcast they want to enable broadcasting
```
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply src/dsl/word/emit/UDP.hpp:145]: Yeah but it's their problem now!

**Comment by TrentHouliston** [comment]
Temporerally going to disable windows requirements as it's having trouble with catch multithreading issues which should be fixed in #70

**Review by Bidski — APPROVED** [review]
LGTM

**Pull request merged**